presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -401,6 +417,27 @@ private void processNewSources()
                 }
             }
 
+            for (int index = operators.size() - 1; index >= 0; index--) {
+                if (operators.get(index).isFinished()) {
```

Comment:
Consider rephrasing this `loop` so that reader doesn't have to think what happens to `index` when `this.operators` is changed.
```
index = findFirstFinishedOperator()
if (found) { // this can be index >= 0 or index.isPresent()
   ....
}
```